### PR TITLE
connect: add MinimumLeafTTL to provider interface

### PR DESCRIPTION
**Pull request by tradel**
This PR adds a method called `MinimumLeafTTL()` to the `Provider` interface as envisioned in RFC for CA external providers. CA providers must return the shortest validity period they are capable of putting into a certificate.

### CODE REVIEW

File: agent/consul/leader_connect.go [hunk 1]
```diff
@@ -224,10 +224,21 @@ func (s *Server) initializeRootCA(provider ca.Provider, conf *structs.CAConfigur
...
 	commonConfig, err := conf.GetCommonConfig()
 	if err != nil {
 		return err
 	}
+	if commonConfig.LeafCertTTL.Seconds() == 0 {
+		commonConfig.LeafCertTTL = provider.MinimumLeafTTL()
+		s.logger.Printf("[WARN] leaf TTL not configured, setting to provider minimum of %v",
+			provider.MinimumLeafTTL())
+	} else if commonConfig.LeafCertTTL < provider.MinimumLeafTTL() {
+		return fmt.Errorf("configured leaf TTL of %v is less than provider minimum of %v",
```
Comment (i0rek) [inline agent/consul/leader_connect.go]: What happens if this fails? Could we just bump the leaf TTL up so that it matches the provider one?

Comment (tradel) [reply agent/consul/leader_connect.go]: According to the RFC, we should fail to configure the CA, unless the `ForceWithoutCrossSigning` parameter has been set to `true`. That will be implemented in a following PR.

File: agent/connect/ca/provider_vault.go [hunk 2]
```diff
@@ -369,6 +370,11 @@ func (v *VaultProvider) CrossSignCA(cert *x509.Certificate) (string, error) {
 	return xcCert, nil
 }
 
+// MinimumLeafTTL returns the shortest validity period supported by this provider.
+func (v *VaultProvider) MinimumLeafTTL() time.Duration {
+	return 1 * time.Hour
```
[the remark this answers is not in the document]
Comment (tradel) [reply agent/connect/ca/provider_vault.go]: I suggest it's more readable this way, and this is the standard we seem to be following across the codebase (e.g. https://github.com/hashicorp/consul/blob/c10f83ebee52e4bcec50a99539a2e540ac7eafed/agent/consul/config.go#L507-L511).

File: agent/consul/leader_connect.go [hunk 3]
```diff
@@ -224,10 +224,21 @@ func (s *Server) initializeRootCA(provider ca.Provider, conf *structs.CAConfigur
...
+	// Check if the configured leaf cert TTL would be less than allowed by the provider.
+	// If so, abort with an error. Also, grab the configured PK type for later use when
+	// creating CSRs.
 	commonConfig, err := conf.GetCommonConfig()
 	if err != nil {
 		return err
 	}
+	if commonConfig.LeafCertTTL.Seconds() == 0 {
+		commonConfig.LeafCertTTL = provider.MinimumLeafTTL()
+		s.logger.Printf("[WARN] leaf TTL not configured, setting to provider minimum of %v",
```
[the remark this answers is not in the document]
Comment (tradel) [reply agent/consul/leader_connect.go]: @banks felt (in previous feedback) that it was important enough to merit a warning.

**Comment by banks** [comment]
@tradel can you rebase and resolve those conflicts with master? I think this was pretty close from last time but I can take another look once the merge is clean.

**Comment by banks** [comment]
Rebased and cleaned up some Protoc merge conflicts. Will do a final pass reviewing and testing...

File: vendor/modules.txt [hunk 4]
```diff
@@ -120,10 +120,18 @@ github.com/go-ole/go-ole/oleutil
...
 github.com/gogo/protobuf/types
 github.com/gogo/protobuf/jsonpb
 github.com/gogo/protobuf/proto
+=======
+github.com/gogo/protobuf/types
+github.com/gogo/protobuf/jsonpb
+github.com/gogo/protobuf/proto
+github.com/gogo/protobuf/sortkeys
+github.com/gogo/protobuf/gogoproto
+>>>>>>> ran update-vendor
```
Comment (banks) [inline vendor/modules.txt]: Ooops... I thought I'd caught all of these.

**Comment by banks** [comment]
On second thoughts, I've decided to ditch this whole idea.

The issue is that we can't validate the MinTTL at config validation time because you need a whole provider instance and that might need to configure and setup some backend to find out the min TTL etc which is all way out of scope for config validation.

But if we hard-fail in the `initializeRootCA` method like this PR does then that will make the entire Consul cluster fail unrecoverably for as long as the CA config is in place as no leader will be able to get past this error which is way to much of a risk for a simple misconfiguration.

The best we can do here would be to _not enforce this_ min TTL and just let the Signing requests fail with a meaningful error message like "Requested TTL of 1h less than supported minimum of 1 day". At this point the operator can correct the config and we didn't silently downgrade security but we also kept the Consul cluster up for everything else except cert signing.

But given that, this whole extra interface method etc. becomes unnecessary and adds no value - CA providers can just return meaningful errors when they don't support a given configuration at Sign time.

So I'm going to abandon this for now for simplicity.

**Pull request closed without merging**